### PR TITLE
Introduce a `groupby` attribute for correlation computation

### PR DESCRIPTION
**Pull request by Marius1311**
This introduces a `groupby` parameter for `mapping_problem.correlate()` to evaluate Pearson/Spearmann correlation over groups of cells. I've also adapted tests slightly to cover this case.

### CODE REVIEW

File: src/moscot/problems/space/_mixins.py [hunk 1]
```diff
@@ -396,7 +396,9 @@ def correlate(  # type: ignore[misc]
         self: SpatialMappingMixinProtocol[K, B],
         var_names: Optional[Sequence[str]] = None,
         corr_method: Literal["pearson", "spearman"] = "pearson",
-    ) -> Mapping[Tuple[K, K], pd.Series]:
+        device: Optional[Device_t] = None,
+        groupby: Optional[str] = None,
```
Comment (giovp) [inline src/moscot/problems/space/_mixins.py:400]: what about calling it `groups`? My understanding is that this is used to compute correlation between groups, see lines 457, 458. I think `groups` is more fitting and to some extent mirrors the usage of the argument in other scverse packages.

File: src/moscot/problems/space/_mixins.py [hunk 2]
```diff
@@ -411,7 +413,13 @@ def correlate(  # type: ignore[misc]
 
             - ``'pearson'`` - `Pearson correlation <https://en.wikipedia.org/wiki/Pearson_correlation_coefficient>`_.
             - ``'spearman'`` - `Spearman's rank correlation
-              <https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient>`_.
+                <https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient>`_.
+
+        groupby
+            Optional `obs` field in `AnnData` to compute correlations over categorical groups.
+
```
Comment (giovp) [inline src/moscot/problems/space/_mixins.py:420]: remove the line space between arguments

File: src/moscot/problems/space/_mixins.py [hunk 3]
```diff
@@ -432,19 +440,46 @@ def correlate(  # type: ignore[misc]
         if sp.issparse(gexp_sc):
             gexp_sc = gexp_sc.toarray()
 
-        corrs = {}
+        corrs = {}  # type: ignore
```
Comment (giovp) [inline src/moscot/problems/space/_mixins.py:443]:
```suggestion
        corrs: dict[str,Union[pd.Series,dict[str,pd.Series]] = {}
```

Comment (giovp) [reply src/moscot/problems/space/_mixins.py:443]: this might work. Would try to type it correctly

File: src/moscot/problems/space/_mixins.py [hunk 4]
```diff
@@ -432,19 +440,46 @@ def correlate(  # type: ignore[misc]
...
+            # predict spatial feature expression
+            gexp_pred_sp = val.to(device=device).pull(gexp_sc, scale_by_marginals=True)
+
+            # loop over groups and compute correlations
+            for group, group_mask in group_masks.items():
+                if np.sum(group_mask) < 2:
+                    logger.debug(f"Skipping `group={group}` as it contains less then 2 samples.")
+                    continue
+
+                # import pdb; pdb.set_trace()
```
Comment (giovp) [inline src/moscot/problems/space/_mixins.py:475]: remove

**Review by giovp — CHANGES_REQUESTED** [review]
looks good! Test also looks good. Minor comments. Thanks for the PR!

File: src/moscot/problems/space/_mixins.py (hunk 1 above)
Comment (Marius1311) [reply src/moscot/problems/space/_mixins.py:400]: I actually prefer `groupby` - you still get a vector of the same length, i.e. length `n_genes`. The difference to before it that when you compute correlation, you only take into account the cells in the current group. In other words, you now get one such vector (containing correlations) per problem and per group. 

I tried to follow scanpy here a bit, see e.g. the violin plotting function, which uses `grouopby` in a similar fashion. https://scanpy.readthedocs.io/en/stable/generated/scanpy.pl.violin.html

On the other hand, in scanpy's scatter plot function, they use a `groups` attribute to restrict visualization to certain groups, which is not really what we are doing here: https://scanpy.readthedocs.io/en/stable/generated/scanpy.pl.scatter.html

File: src/moscot/problems/space/_mixins.py (hunk 2 above)
Comment (Marius1311) [reply src/moscot/problems/space/_mixins.py:420]: done.

File: src/moscot/problems/space/_mixins.py (hunk 4 above)
Comment (Marius1311) [reply src/moscot/problems/space/_mixins.py:475]: done.

File: src/moscot/problems/space/_mixins.py (hunk 3 above)
Comment (Marius1311) [reply src/moscot/problems/space/_mixins.py]: done.

**Comment by Marius1311** [comment]
Thanks for the review @giovp, incorporated all your suggestions, except for the `groupby` -> `groups` change, see comment above.

**Review by selmanozleyen — APPROVED** [review]
Looks good

**Comment by Marius1311** [comment]
Great, thanks @selmanozleyen @giovp, merging this.

**Comment by selmanozleyen** [comment]
@giovp  I think the problem with linting is here for sure. I thought maybe it's a link checking error because of connection but it fails only after this branch (I rerun the checks on older PR's)

**Comment by giovp** [comment]
yeah I think the problem is something in the docs of the `correlate` method, not a linkcheck

**Comment by selmanozleyen** [comment]
It's from the warning. Since the `-W` flag is set it gives error when warning is issued. I removed the flag and it works locally https://www.sphinx-doc.org/en/master/man/sphinx-build.html

**Pull request merged**